src/main/java/com/cst438/controllers/GradeBookController.java

```diff
@@ -0,0 +1,262 @@
+		System.out.println("Gradebook - calcFinalGrades for course " + course_id);
+		
+		// check that this request is from the course instructor 
+		String email = "dwisneski@csumb.edu";  // user name (should be instructor's email) 
+		
+		Course c = courseRepository.findById(course_id).orElse(null);
+		if (!c.getInstructor().equals(email)) {
+			throw new ResponseStatusException( HttpStatus.UNAUTHORIZED, "Not Authorized. " );
+		}
+		
+		CourseDTOG cdto = new CourseDTOG();
+		cdto.course_id = course_id;
+		cdto.grades = new ArrayList<>();
+		for (Enrollment e: c.getEnrollments()) {
+			double total=0.0;
+			int count = 0;
+			for (AssignmentGrade ag : e.getAssignmentGrades()) {
+				count++;
+				total = total + Double.parseDouble(ag.getScore());
+			}
+			double average = total/count;
+			CourseDTOG.GradeDTO gdto = new CourseDTOG.GradeDTO();
+			gdto.grade=letterGrade(average);
+			gdto.student_email=e.getStudentEmail();
+			gdto.student_name=e.getStudentName();
+			cdto.grades.add(gdto);
+			System.out.println("Course="+course_id+" Student="+e.getStudentEmail()+" grade="+gdto.grade);
+		}
+		
+		registrationService.sendFinalGrades(course_id, cdto);
+	}
+	
+	private String letterGrade(double grade) {
+		if (grade >= 90) return "A";
+		if (grade >= 80) return "B";
+		if (grade >= 70) return "C";
+		if (grade >= 60) return "D";
+		return "F";
+	}
+	
+	@PutMapping("/gradebook/{id}")
+	@Transactional
+	public void updateGradebook (@RequestBody GradebookDTO gradebook, @PathVariable("id") Integer assignmentId ) {
+		
+		String email = "dwisneski@csumb.edu";  // user name (should be instructor's email) 
+		checkAssignment(assignmentId, email);  // check that user name matches instructor email of the course.
+		
+		// for each grade in gradebook, update the assignment grade in database 
+		System.out.printf("%d %s %d\n",  gradebook.assignmentId, gradebook.assignmentName, gradebook.grades.size());
+		
+		for (GradebookDTO.Grade g : gradebook.grades) {
+			System.out.printf("%s\n", g.toString());
+			AssignmentGrade ag = assignmentGradeRepository.findById(g.assignmentGradeId).orElse(null);
+			if (ag == null) {
+				throw new ResponseStatusException( HttpStatus.BAD_REQUEST, "Invalid grade primary key. "+g.assignmentGradeId);
+			}
+			ag.setScore(g.grade);
+			System.out.printf("%s\n", ag.toString());
+			
+			assignmentGradeRepository.save(ag);
+		}
+		
+	}
+	
+	private Assignment checkAssignment(int assignmentId, String email) {
+		// get assignment 
+		Assignment assignment = assignmentRepository.findById(assignmentId).orElse(null);
+		if (assignment == null) {
+			throw new ResponseStatusException( HttpStatus.BAD_REQUEST, "Assignment not found. "+assignmentId );
+		}
+		// check that user is the course instructor
+		if (!assignment.getCourse().getInstructor().equals(email)) {
+			throw new ResponseStatusException( HttpStatus.UNAUTHORIZED, "Not Authorized. " );
+		}
+		
+		return assignment;
+	}
+	
+	@PostMapping("/course/{course_id}/assignments")
```

Comment:
Optional, but it could be good to separate out the Assignment mappings into an AssignmentController, rather than having addAssignment and deleteAssignment contained in the GradeBookController